File: pet_vaccine_tracker.js
// pet_vaccine_tracker.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');
const chalk = require('chalk');

const DATA_FILE = 'pets.json';

class Vaccine {
    constructor(name, date, validMonths) {
        this.name = name;
        this.date = new Date(date);
        this.validMonths = validMonths;
    }

    expiryDate() {
        const d = new Date(this.date);
        d.setMonth(d.getMonth() + this.validMonths);
        return d;
    }

    daysUntilExpiry() {
        const diff = this.expiryDate() - new Date();
        return Math.ceil(diff / (1000 * 60 * 60 * 24));
    }

    isExpired() { return this.daysUntilExpiry() < 0; }

    toJSON() {
        return {
            name: this.name,
            date: this.date.toISOString().slice(0,10),
            validMonths: this.validMonths
        };
    }

    static fromJSON(d) {
        return new Vaccine(d.name, d.date, d.validMonths);
    }
}

class Pet {
    constructor(name, species, birthdate, owner, id = null) {
        this.id = id;
        this.name = name;
        this.species = species;
        this.birthdate = new Date(birthdate);
        this.owner = owner;
        this.vaccines = [];
    }

    toJSON() {
        return {
            id: this.id,
            name: this.name,
            species: this.species,
            birthdate: this.birthdate.toISOString().slice(0,10),
            owner: this.owner,
            vaccines: this.vaccines.map(v => v.toJSON())
        };
    }

    static fromJSON(d) {
        const p = new Pet(d.name, d.species, d.birthdate, d.owner, d.id);
        p.vaccines = (d.vaccines || []).map(v => Vaccine.fromJSON(v));
        return p;
    }
}

class Tracker {
    constructor() {
        this.pets = [];
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            const data = JSON.parse(fs.readFileSync(DATA_FILE));
            this.pets = data.map(d => Pet.fromJSON(d));
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.pets.map(p => p.toJSON()), null, 2));
    }

    nextId() {
        return this.pets.reduce((max, p) => Math.max(max, p.id || 0), 0) + 1;
    }

    getPet(id) {
        return this.pets.find(p => p.id === id);
    }

    addPet(name, species, birthdate, owner) {
        const pet = new Pet(name, species, birthdate, owner, this.nextId());
        this.pets.push(pet);
        this.save();
        console.log(`✅ Pet added: ${name} (ID: ${pet.id})`);
    }

    addVaccine(petId, vaccineName, date, validMonths) {
        const pet = this.getPet(petId);
        if (!pet) {
            console.log(`❌ Pet with ID ${petId} not found.`);
            return;
        }
        const v = new Vaccine(vaccineName, date, validMonths);
        pet.vaccines.push(v);
        this.save();
        console.log(`💉 Vaccine '${vaccineName}' added for ${pet.name}.`);
    }

    listPets() {
        if (this.pets.length === 0) {
            console.log('No pets.');
            return;
        }
        console.log('\n🐾 Pets:');
        for (const p of this.pets) {
            console.log(`  [${p.id}] ${p.name} (${p.species}) – Owner: ${p.owner}`);
        }
    }

    showPet(petId) {
        const pet = this.getPet(petId);
        if (!pet) {
            console.log(`❌ Pet with ID ${petId} not found.`);
            return;
        }
        console.log(`\n🐶 ${chalk.cyan(pet.name)} (${pet.species})`);
        console.log(`  Owner: ${pet.owner}`);
        console.log(`  Birth: ${pet.birthdate.toISOString().slice(0,10)}`);
        if (pet.vaccines.length === 0) {
            console.log('  No vaccines recorded.');
            return;
        }
        console.log('  Vaccines:');
        for (let i = 0; i < pet.vaccines.length; i++) {
            const v = pet.vaccines[i];
            const days = v.daysUntilExpiry();
            let status;
            if (days < 0) {
                status = chalk.red(`OVERDUE by ${-days} days 🚨`);
            } else if (days <= 30) {
                status = chalk.yellow(`expires in ${days} days ⚠️`);
            } else {
                status = chalk.green(`expires in ${days} days`);
            }
            console.log(`    [${i+1}] ${v.name} – ${v.date.toISOString().slice(0,10)} – ${status}`);
        }
    }

    checkExpiring(daysThreshold = 30) {
        const expiring = [];
        for (const p of this.pets) {
            for (const v of p.vaccines) {
                const days = v.daysUntilExpiry();
                if (days >= 0 && days <= daysThreshold) {
                    expiring.push({ pet: p, vaccine: v, days });
                } else if (days < 0) {
                    expiring.push({ pet: p, vaccine: v, days });
                }
            }
        }
        if (expiring.length === 0) {
            console.log('✅ No vaccines expiring soon.');
            return;
        }
        console.log(`\n⏰ Vaccines expiring within ${daysThreshold} days (or overdue):`);
        for (const e of expiring) {
            const status = e.days < 0 ? chalk.red(`OVERDUE by ${-e.days} days`) : chalk.yellow(`${e.days} days left`);
            console.log(`  ${e.pet.name} – ${e.vaccine.name} – ${status}`);
        }
    }
}

program
    .command('add-pet <name> <species> <birthdate> <owner>')
    .action((name, species, birthdate, owner) => {
        const tracker = new Tracker();
        tracker.addPet(name, species, birthdate, owner);
    });

program
    .command('add-vaccine <petId> <vaccineName> <date> <validMonths>')
    .action((petId, vaccineName, date, validMonths) => {
        const tracker = new Tracker();
        tracker.addVaccine(parseInt(petId), vaccineName, date, parseInt(validMonths));
    });

program
    .command('list')
    .action(() => {
        const tracker = new Tracker();
        tracker.listPets();
    });

program
    .command('show <petId>')
    .action((petId) => {
        const tracker = new Tracker();
        tracker.showPet(parseInt(petId));
    });

program
    .command('check')
    .option('-d, --days <n>', 'Days threshold', parseInt, 30)
    .action((options) => {
        const tracker = new Tracker();
        tracker.checkExpiring(options.days);
    });

program.parse(process.argv);

File: pet_vaccine_tracker.rb
# pet_vaccine_tracker.rb
require 'json'
require 'date'
require 'colorize'
require 'optparse'

DATA_FILE = 'pets.json'

class Vaccine
  attr_accessor :name, :date, :valid_months

  def initialize(name, date, valid_months)
    @name = name
    @date = Date.parse(date)
    @valid_months = valid_months
  end

  def expiry_date
    @date >> valid_months
  end

  def days_until_expiry
    (expiry_date - Date.today).to_i
  end

  def expired?
    days_until_expiry < 0
  end

  def to_hash
    { name: @name, date: @date.iso8601, valid_months: @valid_months }
  end

  def self.from_hash(h)
    new(h['name'], h['date'], h['valid_months'])
  end
end

class Pet
  attr_accessor :id, :name, :species, :birthdate, :owner, :vaccines

  def initialize(name, species, birthdate, owner, id = nil)
    @id = id
    @name = name
    @species = species
    @birthdate = Date.parse(birthdate)
    @owner = owner
    @vaccines = []
  end

  def to_hash
    {
      id: @id,
      name: @name,
      species: @species,
      birthdate: @birthdate.iso8601,
      owner: @owner,
      vaccines: @vaccines.map(&:to_hash)
    }
  end

  def self.from_hash(h)
    p = new(h['name'], h['species'], h['birthdate'], h['owner'], h['id'])
    p.vaccines = h['vaccines'].map { |v| Vaccine.from_hash(v) }
    p
  end
end

class Tracker
  attr_reader :pets

  def initialize
    @pets = []
    load
  end

  def load
    return unless File.exist?(DATA_FILE)
    data = JSON.parse(File.read(DATA_FILE))
    @pets = data.map { |h| Pet.from_hash(h) }
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@pets.map(&:to_hash)))
  end

  def next_id
    @pets.map(&:id).max.to_i + 1
  end

  def get_pet(id)
    @pets.find { |p| p.id == id }
  end

  def add_pet(name, species, birthdate, owner)
    pet = Pet.new(name, species, birthdate, owner, next_id)
    @pets << pet
    save
    puts "✅ Pet added: #{name} (ID: #{pet.id})"
  end

  def add_vaccine(pet_id, vaccine_name, date, valid_months)
    pet = get_pet(pet_id)
    unless pet
      puts "❌ Pet with ID #{pet_id} not found."
      return
    end
    v = Vaccine.new(vaccine_name, date, valid_months)
    pet.vaccines << v
    save
    puts "💉 Vaccine '#{vaccine_name}' added for #{pet.name}."
  end

  def list_pets
    if @pets.empty?
      puts "No pets."
      return
    end
    puts "\n🐾 Pets:"
    @pets.each do |p|
      puts "  [#{p.id}] #{p.name} (#{p.species}) – Owner: #{p.owner}"
    end
  end

  def show_pet(pet_id)
    pet = get_pet(pet_id)
    unless pet
      puts "❌ Pet with ID #{pet_id} not found."
      return
    end
    puts "\n🐶 #{pet.name.colorize(:cyan)} (#{pet.species})"
    puts "  Owner: #{pet.owner}"
    puts "  Birth: #{pet.birthdate.iso8601}"
    if pet.vaccines.empty?
      puts "  No vaccines recorded."
      return
    end
    puts "  Vaccines:"
    pet.vaccines.each_with_index do |v, i|
      days = v.days_until_expiry
      status = if days < 0
                 "OVERDUE by #{-days} days 🚨".red
               elsif days <= 30
                 "expires in #{days} days ⚠️".yellow
               else
                 "expires in #{days} days".green
               end
      puts "    [#{i+1}] #{v.name} – #{v.date.iso8601} – #{status}"
    end
  end

  def check_expiring(days_threshold = 30)
    expiring = []
    @pets.each do |p|
      p.vaccines.each do |v|
        days = v.days_until_expiry
        if days >= 0 && days <= days_threshold
          expiring << { pet: p, vaccine: v, days: days }
        elsif days < 0
          expiring << { pet: p, vaccine: v, days: days }
        end
      end
    end
    if expiring.empty?
      puts "✅ No vaccines expiring soon."
      return
    end
    puts "\n⏰ Vaccines expiring within #{days_threshold} days (or overdue):"
    expiring.each do |e|
      status = e[:days] < 0 ? "OVERDUE by #{-e[:days]} days".red : "#{e[:days]} days left".yellow
      puts "  #{e[:pet].name} – #{e[:vaccine].name} – #{status}"
    end
  end
end

options = {}
$command = ARGV.shift
if $command.nil?
  puts "Usage: pet_vaccine_tracker.rb <command> [options]"
  exit 1
end

app = Tracker.new

case $command
when "add-pet"
  name, species, birthdate, owner = ARGV.shift(4)
  app.add_pet(name, species, birthdate, owner)
when "add-vaccine"
  pet_id, vaccine_name, date, valid_months = ARGV.shift(4)
  app.add_vaccine(pet_id.to_i, vaccine_name, date, valid_months.to_i)
when "list"
  app.list_pets
when "show"
  pet_id = ARGV.shift.to_i
  app.show_pet(pet_id)
when "check"
  days = 30
  if ARGV.include?('--days')
    idx = ARGV.index('--days')
    days = ARGV[idx+1].to_i if idx
  end
  app.check_expiring(days)
else
  puts "Unknown command."
end
